File: src/main/java/com/example/travelator/actiontocalculation/version4/InMemoryTrips.java
package com.example.travelator.actiontocalculation.version4;

import java.time.Clock;
import java.time.Instant;
import java.util.*;

import static java.util.stream.Collectors.toSet;

public class InMemoryTrips implements Trips {
	private final Map<String, Set<Trip>> trips;
	private final Clock clock;
	
	public InMemoryTrips(Clock clock) {
		this.trips = new HashMap<>();
		this.clock = clock;
	}
	
	void addTrip(Trip trip) {
		var existingTrips = trips.getOrDefault(trip.getCustomerId(), new LinkedHashSet<>());
		existingTrips.add(trip);
		trips.putIfAbsent(trip.getCustomerId(), existingTrips);
//		trips.computeIfAbsent(trip.getCustomerId(), (id) -> new LinkedHashSet<>()).add(trip);
	}
	
	@Override
	public Set<Trip> currentTripsFor(String customerId, Instant at) {
		return tripsFor(customerId).stream()
			.filter(trip -> trip.isPlannedToBeActiveAt(at))
			.collect(toSet());
	}
	
	private Collection<Trip> tripsFor(String customerId) {
		return trips.getOrDefault(customerId, Collections.emptySet());
	}
}

File: src/main/java/com/example/travelator/actiontocalculation/version5/InMemoryTrips.kt
package com.example.travelator.actiontocalculation.version5

import java.time.Clock
import java.time.Instant
import java.util.stream.Collectors

class InMemoryTrips(clock: Clock) : Trips {
    private val trips: MutableMap<String, MutableSet<Trip>>
    private val clock: Clock

    init {
        trips = HashMap()
        this.clock = clock
    }

    fun addTrip(trip: Trip) {
        val existingTrips = trips.getOrDefault(trip.customerId, LinkedHashSet())
        existingTrips.add(trip)
        trips.putIfAbsent(trip.customerId, existingTrips)
        //		trips.computeIfAbsent(trip.getCustomerId(), (id) -> new LinkedHashSet<>()).add(trip);
    }

    override fun currentTripsFor(customerId: String, at: Instant): Set<Trip> {
        return tripsFor(customerId).stream()
            .filter { trip: Trip -> trip.isPlannedToBeActiveAt(at) }
            .collect(Collectors.toSet())
    }

    private fun tripsFor(customerId: String): Collection<Trip> {
        return trips.getOrDefault(customerId, emptySet())
    }
}